File: app/src/main/java/com/example/bridgeapp/structure/SuitObject.java
package com.example.bridgeapp.structure;

import android.os.Parcel;
import android.os.Parcelable;

import com.example.bridgeapp.util.CardSuit;
import com.example.bridgeapp.util.CardValue;

import java.util.Arrays;

public class SuitObject implements Parcelable {
    private final CardSuit suit;
    private final CardValue[] cards;

    public static final Creator<SuitObject> CREATOR = new Creator<SuitObject>() {
        @Override
        public SuitObject createFromParcel(Parcel in) {
            return new SuitObject(in);
        }

        @Override
        public SuitObject[] newArray(int size) {
            return new SuitObject[size];
        }
    };

    public SuitObject(CardSuit suit) {
        this.suit = suit;
        cards = new CardValue[CardValue.values().length];
    }

    public void addCard(CardValue card){
        if(!hasCard(card)){
            cards[card.ordinal()] = card;
        }
    }

    public boolean hasCard(CardValue card){
        return Arrays.asList(cards).contains(card);
    }

    public CardSuit getSuit() {
        return suit;
    }

    public CardValue[] getCards() {
        return cards;
    }

//Parcelable implementation

    protected SuitObject(Parcel in) {
        this.suit = CardSuit.valueOf(in.readString());
        String[] cardNames = in.createStringArray();

        cards = new CardValue[cardNames.length];

        for(int i = 0; i < cardNames.length; i++) {
            if (cardNames[i] != null)
                cards[i] = CardValue.valueOf(cardNames[i]);
        }
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        String[] cardNames = new String[cards.length];
        for(int i = 0; i < cardNames.length; i++) {
            if(cards[i] != null)
                cardNames[i] = cards[i].name();
        }

        dest.writeString(this.suit.name());
        dest.writeStringArray(cardNames);
    }
}

File: app/src/main/java/com/example/bridgeapp/structure/SuitObject.kt
package com.example.bridgeapp.structure

import android.os.Parcel
import android.os.Parcelable
import com.example.bridgeapp.util.CardSuit
import com.example.bridgeapp.util.CardValue

class SuitObject : Parcelable {
    val suit: CardSuit
    val cards: Array<CardValue?>

    constructor(suit: CardSuit) {
        this.suit = suit
        cards = arrayOfNulls(CardValue.values().size)
    }

    fun addCard(card: CardValue) {
        if (!hasCard(card)) {
            cards[card.ordinal] = card
        }
    }

    fun hasCard(card: CardValue?): Boolean {
        return listOf(*cards).contains(card)
    }

    //Parcelable implementation
    private constructor(`in`: Parcel) {
        suit = CardSuit.valueOf(`in`.readString()!!)
        val cardNames = `in`.createStringArray()
        cards = arrayOfNulls(cardNames!!.size)
        for (i in cardNames.indices) {
            if (cardNames[i] != null) cards[i] = CardValue.valueOf(cardNames[i]!!)
        }
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        val cardNames = arrayOfNulls<String>(cards.size)
        for (i in cardNames.indices) {
            if (cards[i] != null) cardNames[i] = cards[i]!!.name
        }
        dest.writeString(suit.name)
        dest.writeStringArray(cardNames)
    }

    companion object {
        @JvmField val CREATOR: Parcelable.Creator<SuitObject?> = object : Parcelable.Creator<SuitObject?> {
            override fun createFromParcel(`in`: Parcel): SuitObject? {
                return SuitObject(`in`)
            }

            override fun newArray(size: Int): Array<SuitObject?> {
                return arrayOfNulls(size)
            }
        }
    }
}